openvdb_points/tools/IndexFilter.h

```diff
@@ -0,0 +1,301 @@
+// of its contributors may be used to endorse or promote products derived
+// from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+// IN NO EVENT SHALL THE COPYRIGHT HOLDERS' AND CONTRIBUTORS' AGGREGATE
+// LIABILITY FOR ALL CLAIMS REGARDLESS OF THEIR BASIS EXCEED US$250.00.
+//
+///////////////////////////////////////////////////////////////////////////
+//
+/// @file IndexFilter.h
+///
+/// @authors Dan Bailey
+///
+/// @brief  Index filters primarily designed to be used with a FilterIndexIter.
+///
+
+
+#ifndef OPENVDB_TOOLS_INDEX_FILTER_HAS_BEEN_INCLUDED
+#define OPENVDB_TOOLS_INDEX_FILTER_HAS_BEEN_INCLUDED
+
+#include <openvdb/version.h>
+#include <openvdb/Types.h>
+
+#include <openvdb/math/Transform.h>
+
+#include <openvdb_points/tools/IndexIterator.h>
+#include <openvdb_points/tools/AttributeArray.h>
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace tools {
+
+
+////////////////////////////////////////
+
+
+namespace index_filter_internal {
+
+template <typename RandGenT, typename IntType>
+struct RandGen
+{
+    RandGen(const unsigned int seed, const IntType length) : randGen(seed, 0, length-1) { }
+    IntType operator()(IntType n) { return randGen(); }
+    openvdb::math::RandInt<IntType, RandGenT> randGen;
+};
+
+// generate a random subset of n indices from the range [0:m]
+template <typename RandGenT, typename IntType>
+void generateRandomSubset(std::vector<IntType>& values, const unsigned int seed, const IntType n, const IntType m)
+{
+    if (n <= 0)     return;
+
+    if (n >= m) {
+        values.reserve(n);
+        for (int i = 0; i < int(m); i++)    values.push_back(IntType(i));
+        return;
+    }
+
+    // reserve and fill vector with ascending indices
+    values.reserve(m);
+    for (int i = 0; i < int(m); i++)    values.push_back(i);
+
+    // shuffle indices using random generator
+    RandGen<RandGenT, IntType> randGen(seed, m);
+    std::random_shuffle(values.begin(), values.end(), randGen);
+
+    // sort the subset of the indices vector that will be used
+    std::sort(values.begin(), values.begin() + n);
```

Comment:
Well, there's a tradeoff here. If the array is resized, that means allocating a new array (of the new size), copying over the values, de-allocating the old array. This all takes time, so this method simply sorts the first n elements of the existing array, then leaves it with elements beyond that being unsorted. The idea is that the user should only ever use the first n elements, as these are the only ones guaranteed to be correct and the assumption is this data is only temporary anyway so the only benefit is marginally improved memory usage.

I did actually have a second method where it inserts random values into a set repeatedly until it hits the required number of values and switch to using this value if n < 10%. Much more efficient for small n, however there was a noticeable difference in the method changing when it crossed this threshold and I couldn't find a good way of eradicating this, so I removed it altogether.